File: data-structure-exercise/二分查找.cpp
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *R;
    int length;
} SSTable;

void Input(SSTable &ST)
{
    printf ("How many number you will input: ");
    scanf ("%d", &ST.length);
    ST.R = (int *) malloc (sizeof(int) * ST.length);
    printf ("Input all number: ");
    for (int i = 1; i <= ST.length; i++)
    {
        scanf ("%d", &ST.R[i]);
    }
}

int Search_bin(SSTable ST, int key)
{
    int low = 1;
    int high = ST.length;
    while (low <= high)
    {
        int mid = (low + high) / 2;
        if (ST.R[mid] == key)
        {
            return mid;
        }
        else if (ST.R[mid] < key)
        {
            low = mid + 1;
        }
        else
        {
            high = mid - 1;
        }
    }
    return 0;
}

int main()
{
    SSTable ST;
    Input(ST);
    int flag = Search_bin(ST, 3);
    if (flag == 0)
    {
        printf ("Not Found\n");
    }
    else
    {
        printf ("Find 3 in %d in the sequence\n", flag);
    }
}

File: python/ICPUP/二分查找.py
def search(L, e):
    def bSearch(L, e, low, high):
        if high == low:
            return L[low] == e
        mid = (low + high) // 2

        if L[mid] == e:
            return True
        elif L[mid] > e:
            if low == mid:
                return False
            else:
                return bSearch(L, e, low, mid - 1)
        else:
            return bSearch(L, e, mid + 1, high)

    if len(L) == 0:
        return False
    else:
        return bSearch(L, e, 0, len(L) - 1)

L = [1, 2, 5, 3, 6, 7, 9, 10]
search(L, 8)
